components/tflite-lib/tensorflow/lite/micro/kernels/esp_nn/conv.cc

```diff
@@ -166,13 +172,24 @@ inline void EvalQuantizedPerChannel(
       TFLITE_DCHECK_EQ(bias_shape.FlatSize(), output_depth);
     }
 
-    void *scratch_buf = NULL;// = calloc(1, scratch_buf_size + 128 + 16);
-    //int align_offset = 16 - ((int) (scratch_buf) & 15);
-    //esp_nn_set_conv_scratch_buf((void *) ((int) scratch_buf + align_offset));
+  void *scratch_buf = NULL;
+#if USE_DYNAMIC_ALLOC
+    int scratch_buf_size = esp_nn_get_conv_scratch_size(
+        input_width, input_height, input_depth,
+        output_depth, filter_width, filter_height);
+    scratch_buf = heap_caps_malloc(scratch_buf_size + 16, MALLOC_CAP_SPIRAM | MALLOC_CAP_8BIT);
```

Comment:
Does the **(MALLOC_CAP_SPIRAM | MALLOC_CAP_8BIT)** capability mean from SPIRAM or regular RAM?  

Micropython uses **SPIRAM_USE_MEMMAP=y** so I just want to confirm that this malloc won´t allocate SPRAM out from under micropython.

I hope it means that if **SPIRAM_USE_MALLOC=y** then it would allocate from SPIRAM but otherwise allocate from DRAM.

Reply:
Hi @mocleiri 

`SPIRAM_USE_MALLOC=y` makes SPIRAM allocatable using `malloc` as well.

`MALLOC_CAP_SPIRAM` capability does mean we are explicitly asking for memory to be allocatable in SPIRAM if available.
If you have chosen `SPIRAM_USE_CAPS_ALLOC` you should use `MALLOC_CAP_INTERNAL` capability for internal and `MALLOC_CAP_SPIRAM` for external memory using heap_caps.

If you have enabled,`SPIRAM_USE_MEMMAP` option however, SPIRAM is enabled, however it is not available for allocation using `MALLOC_CAP_SPIRAM` caps or `malloc`.
So, in your case, the above allocation will go to the DRAM.

Reply:
Thanks for clarifying it.  

For the ESP32-CAM firmware we run in SPIRAM_USE_MALLOC=y mode and split the 4MB PSRAM in half.  2MB for micropython heap and 2MB for esp32-camera and other uses like this line you added here.

But for the regular boards either there is no PSRAM or micropython uses all of the heap itself using SPIRAM_USE_MEMMAP.

For the tensorflow arena we allocate a byte array from micropython (allocated from its heap in PSRAM) and then pass that through for tflm to use.

If this buffer can get large I may need to make a way to pass through a micropython allocated byte array for this buffer as well on my side.

I don't think you need to anything I'll keep an eye on in and if needed file a pull request for what ever I needed to do to pass through a different buffer.

Reply:
@mocleiri
This bug fix has solved my previous problem of using my custom model trained on edge impulse.
But it seems some errors when I used the model on your https://github.com/mocleiri/tensorflow-micropython-examples/tree/feature/9-add-accelerated-esp-tflm-kernels.
Any advice on config SPIRAM_USE_MEMMAP and heap_caps_malloc?Thanks.